prusti-specs/src/lib.rs

```diff
@@ -21,34 +26,97 @@ macro_rules! handle_result {
     };
 }
 
-pub fn requires(attr: TokenStream, tokens: TokenStream) -> TokenStream {
-    let item: syn::ItemFn = handle_result!(syn::parse2(tokens));
+/// Rewrite an item as required by *all* its specification attributes.
+///
+/// The outer attribute needs to be passed via `attr_kind` and `attr` because the compiler will
+/// detect and execute it as as a procedural macro attribute.
+pub fn rewrite_prusti_attributes(
+    outer_attr_kind: SpecAttributeKind,
+    outer_attr_tokens: TokenStream,
+    item_tokens: TokenStream
+) -> TokenStream {
+    let mut item: syn::ItemFn = handle_result!(syn::parse2(item_tokens));
+
+    // Start with the outer attribute
+    let mut prusti_attributes = vec![
+        (outer_attr_kind, outer_attr_tokens)
+    ];
+
+    // Collect the remaining Prusti attributes, removing them from `item`.
+    let remaining_prusti_attributes = item.attrs.drain_filter(
+        |attr|
+            attr.path.segments.len() == 1
+                && SpecAttributeKind::try_from(attr.path.segments[0].ident.to_string()).is_ok()
```

Comment:
Does this not result in a (maybe unimportant) difference in behaviour? Originally Prusti code would only be triggered from its own attributes. What if I only imported `requires` from Prusti, but had `pure` from a different crate? Here the assumption is that attributes that have the same name as Prusti attributes are definitely the ones from Prusti.

Reply:
Indeed, what you describe might happen. We haven't found a better solution. The correct way would be to let the compiler resolve and execute each procedural macro individually, but that makes the compiler lose all span information (bug https://github.com/rust-lang/rust/issues/43081, open since 2017), which is not acceptable for us.

Reply:
Another case (by Vytautas) which we would not handle  is the following:

```rust
use prusti_contracts::requires as pre;

#[pre(false)]
fn test() { ... }
````

Reply:
Could we require all attributes after the first Prusti one to be valid and trigger errors if we don't recognise them?

Reply:
That's a good idea!

Reply:
Unfortunately, we don't  have the information of which attributes come after or before a procedural macro attribute. Disallowing any other kind of attributes (including e.g. `#[allow(..)]` and doc comments) would be too restrictive.

Reply:
That's a pity. I assumed they would be processed "left to right", i.e. the first proc macro attribute that is processed would only see the attributes that follow it.

Reply:
I thought the same, but apparently all past attributes are still in the `item.attrs` field.